sktime/transformations/series/temporian.py

```diff
@@ -0,0 +1,115 @@
+
+"""Implementation of TemporianTransformer. Contributed by the Temporian team."""
+
+__author__ = ["ianspektor"]  # TODO: add other Temporian devs who contributed to this
+
+from sktime.transformations.base import BaseTransformer
+
+
+# TODO: add usage example to docstring
+# see https://www.sktime.net/en/latest/developer_guide/dependencies.html#dependencies
+# for how to handle soft deps (temporian) there (need to not run doc tests on it)
+class TemporianTransformer(BaseTransformer):
+    """Applies a Temporian function to the input time series.
+
+    This transformer applies a [Temporian](https://temporian.readthedocs.io/en/stable/)
+    function to the input time series.
+
+    The conversion from sktime's internal representation to Temporian's
+    [EventSet](https://temporian.readthedocs.io/en/stable/reference/temporian/EventSet/)
+    and back is handled automatically by the transformer.
+
+    Parameters
+    ----------
+    function : Callable[[temporian.EventSet], temporian.EventSet]
+        Temporian function to apply to the input time series. The function must receive
+        and return a single Temporian EventSet, and can apply an arbitrary number of
+        Temporian operators to its input.
+    """
+
+    _tags = {
+        "univariate-only": False,
+        "authors": [
+            "ianspektor"
+        ],  # TODO: add other Temporian devs who contributed to this
+        "maintainers": [
+            "ianspektor"
+        ],  # TODO: add other Temporian devs who will maintain this
+        "scitype:transform-input": "Series",
+        "scitype:transform-output": "Series",
+        "scitype:instancewise": True,
+        "scitype:transform-labels": "None",
+        # TODO: extend to pd-multiindex and pd_multiindex_hier
+        "X_inner_mtype": "pd.DataFrame",
+        "fit_is_empty": True,
+        "capability:inverse_transform": False,
+        "capability:unequal_length": True,
+        "handles-missing-data": False,
+        "python_dependencies": ["temporian"],
+        "python_version": "<3.12",
+    }
+
+    def __init__(self, function):
+        self.function = function
+
+        super().__init__()
+
+        # TODO: ensure function receives a single EventSet/param with inspect module?
+        #       or is failing in runtime OK?
+        # TODO: @tp.compile the function
+
+    def _transform(self, X, y=None):
+        """Transform X and return a transformed version.
+
+        private _transform containing core logic, called from transform
+
+        Parameters
+        ----------
+        X : pd.DataFrame
+            if self.get_tag("univariate-only")==True:
+                guaranteed to have a single column
+            if self.get_tag("univariate-only")==False: no restrictions apply
+        y : None, present only for interface compatibility
+
+        Returns
+        -------
+        transformed version of X
+        """
+        import temporian as tp
+
+        timestamps_col = X.index.name
```

Comment:
would this not break in a case of a nameless index?

Reply:
I need more info about sktime's internal representations and about exactly what kinds of objects we may be receiving here, you're most likely right, will test thoroughly

Reply:
This is a good point.

@fkiraly, are there specificities on the dataframe we can expect here? For instance, are timestamps always the index, or are they in a column with a specific name? Is the timestamps restricted to some specific dtypes? Can we have a multi-index? Etc.

@ianspektor, what about adding many assert / checks to restrict strongly what you think the code supports. Franz can then tell whether such and such restrictions make sense. It is probably easier to over restrict the supported inputs, rather than debugging failures of a non expected input type.

Reply:
@achom, the specifications are laid out in this notebook: https://github.com/sktime/sktime/blob/main/examples/AA_datatypes_and_datasets.ipynb

From the dev meeting with @ianspektor, a while ago, we've thought it would be best to rely only on the `pd.DataFrame` based types internally. In those, you can rely on the last level of the `MultiIndex` (rows) being time, and ordered monotonically.

Reply:
@fkiraly +1 to @achoum's first question. I thought specifying a single `X_inner_mtype` would mean we always receive the same data structure, but that doesn't seem to be the case.

For example, when running `check_estimator(TemporianTransformer, tests_to_run="test_fit_transform_output")`, `_transform` always receives a `DataFrame`, but the index is sometimes a `DatetimeIndex` and sometimes an `Index`. And, when I run the `test_flat_univariate` in this PR, it's a `PeriodIndex`.

Could you point me towards docs that explain exactly what `_transform` could receive given my tags specification?

Thanks!